app/jobs/async_batch_destroy_job.rb

```diff
@@ -0,0 +1,13 @@
+class AsyncBatchDestroyJob < ActiveJob::Base
+  queue_as :batch_actions
+
+  BATCH_SIZE = 1000
+
+  def perform(model_class, sql_query)
+    begin
+      scoped_records = model_class.constantize.find_by_sql(sql_query + " LIMIT #{BATCH_SIZE}")
+      scoped_records.each { |record| record.destroy! }
```

Comment:
I'm no sure if it will work. If I understand this correctly the `find_by_sql ( + LIMIT 1000)` will find only first 1000 records and no more. For such cases is useful this:

http://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each

It would be nice to have one test on destroy/update more then 1000 records.

But `find_each` works only with `ActiveRecord::Relation` the original `scoped_collection_records` is an instance of ActiveRecord::Relation.

It instead of sql_query use Relation, then AsyncBatchDestroyJob and AsyncBatchUpdateJob will be match simpler.

```ruby
scoped_collection_records.find_each(batch_size: 1000) do |r|
  r.destroy!
end
```

Reply:
Code works fine: we take first 1000 records, destroy it, then the same sql query will grab next 1000 records (because first batch already doesn't exist), destroy it, etc.
I tested it by lowering BATCH_SIZE constant to 1 or 2. 

About using Relation instead of sql_query.
We can't pass complex rails objects (like AR Relation) to delayed jobs, because those objects can't be serialized and put into job handler string. That's why we are bound to deal with pure sql (I'm not happy about that too).